zenith_utils/src/pq_proto.rs

```diff
@@ -0,0 +1,559 @@
+        if len < 4 {
+            bail!("invalid message length: parsing u32");
+        }
+        let bodylen = len - 4;
+
+        // Read message body
+        let mut body_buf: Vec<u8> = vec![0; bodylen as usize];
+        stream.read_exact(&mut body_buf)?;
+
+        let body = Bytes::from(body_buf);
+
+        // Parse it
+        match tag {
+            b'Q' => Ok(Some(FeMessage::Query(FeQueryMessage { body }))),
+            b'P' => Ok(Some(FeParseMessage::parse(body)?)),
+            b'D' => Ok(Some(FeDescribeMessage::parse(body)?)),
+            b'E' => Ok(Some(FeExecuteMessage::parse(body)?)),
+            b'B' => Ok(Some(FeBindMessage::parse(body)?)),
+            b'C' => Ok(Some(FeCloseMessage::parse(body)?)),
+            b'S' => Ok(Some(FeMessage::Sync)),
+            b'X' => Ok(Some(FeMessage::Terminate)),
+            b'd' => Ok(Some(FeMessage::CopyData(body))),
+            tag => Err(anyhow!("unknown message tag: {},'{:?}'", tag, body)),
+        }
+    }
+}
+
+impl FeStartupMessage {
+    // read startup message from the stream
+    pub fn read(stream: &mut impl std::io::Read) -> anyhow::Result<Option<FeMessage>> {
+        const MAX_STARTUP_PACKET_LENGTH: usize = 10000;
+        const CANCEL_REQUEST_CODE: u32 = (1234 << 16) | 5678;
+        const NEGOTIATE_SSL_CODE: u32 = (1234 << 16) | 5679;
+        const NEGOTIATE_GSS_CODE: u32 = (1234 << 16) | 5680;
+
+        // Read length. If the connection is closed before reading anything (or before
+        // reading 4 bytes, to be precise), return None to indicate that the connection
+        // was closed. This matches the PostgreSQL server's behavior, which avoids noise
+        // in the log if the client opens connection but closes it immediately.
+        let len = match stream.read_u32::<BE>() {
+            Ok(len) => len as usize,
+            Err(e) if e.kind() == io::ErrorKind::UnexpectedEof => return Ok(None),
+            Err(e) => return Err(e.into()),
+        };
+
+        if len < 4 || len > MAX_STARTUP_PACKET_LENGTH {
+            bail!("invalid message length");
+        }
+
+        let version = stream.read_u32::<BE>()?;
+        let kind = match version {
+            CANCEL_REQUEST_CODE => StartupRequestCode::Cancel,
+            NEGOTIATE_SSL_CODE => StartupRequestCode::NegotiateSsl,
+            NEGOTIATE_GSS_CODE => StartupRequestCode::NegotiateGss,
+            _ => StartupRequestCode::Normal,
+        };
+
+        // the rest of startup packet are params
+        let params_len = len - 8;
+        let mut params_bytes = vec![0u8; params_len];
+        stream.read_exact(params_bytes.as_mut())?;
+
+        let params_str = str::from_utf8(&params_bytes).unwrap();
+        let params = params_str.split('\0');
+        // let mut options = false;
+        // let mut timelineid: Option<ZTimelineId> = None;
+        // let mut appname: Option<String> = None;
+        let mut params_hash: HashMap<String, String> = HashMap::new();
+        let mut i = 0;
```

Comment:
It's not really obvious to me what this variable `i` is doing. I'm not sure if it needs a better name, or just a comment explaining what's going on, or what...

Reply:
Consider using `chunks` or `chunks_exact` with n = 2 for this

Reply:
Thanks, `chunks` are indeed much better.